crates/asm/subprotocols/bridge-v1/src/state/operator.rs

```diff
@@ -240,35 +288,67 @@ impl OperatorTable {
         &self.active_operators
     }
 
-    /// Updates the active status for multiple operators, inserts new operators,
-    /// and recalculates the aggregated key.
-    ///
-    /// # Parameters
+    /// Atomically applies membership changes by adding new operators and removing existing ones,
+    /// then recalculates the aggregated key.
     ///
-    /// - `updates` - Slice of (operator_index, is_active) pairs for existing operators
-    /// - `inserts` - Slice of new operator MuSig2 public keys to insert (marked as active by
-    ///   default)
+    /// After recalculating, the new N/N script is appended to `historical_nn_scripts` so the latest
+    /// script remains accessible while older entries continue to support validation for previous
+    /// operator configurations.
     ///
     /// # Processing Order
     ///
-    /// Inserts are processed before updates. If an operator index appears in both parameters,
-    /// the update will override the insert's `is_active` value.
+    /// Additions are processed before removals. If an operator index appears in both parameters,
+    /// the removal will override the addition's `is_active` value.
     ///
     /// # Panics
     ///
     /// Panics if:
-    /// - The updates would result in no active operators
+    /// - The changes would result in no active operators
     /// - Sequential operator insertion fails (bitmap index management error)
     /// - `next_idx` overflows `u32::MAX` when inserting new operators (since operator indices are
     ///   never reused, this limits the total number of unique operators that can ever be registered
     ///   to `u32::MAX` or 4,294,967,295 over the bridge's lifetime)
-    pub fn update_multisig_and_recalc_key(
+    pub fn apply_membership_changes(
         &mut self,
-        updates: &[(OperatorIdx, bool)],
-        inserts: &[EvenPublicKey],
+        add_members: &[EvenPublicKey],
+        remove_members: &[OperatorIdx],
     ) {
-        // Handle inserts first
-        for musig2_pk in inserts {
+        self.add_operators(add_members);
+        self.remove_operators(remove_members);
+
+        if !remove_members.is_empty() || !add_members.is_empty() {
+            self.recalculate_aggregated_key();
+            self.historical_nn_scripts
+                .push(Self::build_nn_script(&self.agg_key));
```

Comment:
This is a smart way to approach it.  Should we be doing more explicit tracking for the n-of-n master pubkey "versions"?